### PR TITLE
multi: support non-interactive collectible sends

### PR DESCRIPTION
**Pull request by jharveyb**
Depends on #159, addresses #99 and #121.

Adds support for non-interactive sends of collectibles by allowing splits bearing collectibles, iff they have an unspendable root and exactly one non-root locator.

Fixes https://github.com/lightninglabs/taro/issues/99 
Fixes https://github.com/lightninglabs/taro/issues/121

### CODE REVIEW

File: commitment/commitment_test.go [hunk 1]
```diff
@@ -473,14 +473,52 @@ func TestSplitCommitment(t *testing.T) {
...
+			err: ErrInvalidSplitLocatorCount,
+		},
+		{
+			name: "collectible split commitment",
+			f: func() (*asset.Asset, *SplitLocator, []*SplitLocator) {
+				input := randAsset(
+					t, genesisCollectible,
+					familyKeyCollectible,
+				)
+				root := &SplitLocator{
```
Comment (Roasbeef) [inline commitment/commitment_test.go:505]: 👍

File: commitment/split.go [hunk 2]
```diff
@@ -151,6 +150,18 @@ func NewSplitCommitment(input *asset.Asset, outPoint wire.OutPoint,
 		return nil, ErrInvalidSplitLocator
 	}
 
+	// To transfer a collectible with a split, the split root must be
```
Comment (Roasbeef) [inline commitment/split.go:153]: 👍

File: vm/vm.go [hunk 3]
```diff
@@ -105,12 +105,10 @@ func matchesAssetParams(newAsset, prevAsset *asset.Asset,
...
 func (vm *Engine) validateSplit() error {
-	// Only `Normal` assets can be split, and the change asset should have
-	// a split commitment root.
+	// The asset type must match for all parts of a split, and the change
+	// asset should have a split commitment root.
 	switch {
-	case vm.newAsset.Type != asset.Normal ||
-		vm.splitAsset.Type != asset.Normal:
-
+	case vm.newAsset.Type != vm.splitAsset.Type:
```
Comment (Roasbeef) [inline vm/vm.go:111]: Love how compact all these changes are, gotta love when a plan nicely falls into place 🤓*


\* hidden bugs may apply

File: tarofreighter/chain_porter.go [hunk 4]
```diff
@@ -702,21 +695,6 @@ func (p *ChainPorter) stateStep(currentPkg sendPackage) (*sendPackage, error) {
 
 		return &currentPkg, nil
 
-	// Alternatively, we'll enter this state when we know we don't actually
-	// need a split at all. In this case, we fully consume an input asset,
-	// so the asset created is the same asset w/ the new script key in
-	// place.
-	case SendStatePreparedComplete:
```
Comment (Roasbeef) [inline tarofreighter/chain_porter.go:703]: We can remove the enum value definition above as well.

File: itest/collectible_split_test.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	"github.com/lightninglabs/taro/tarorpc"
+	"github.com/stretchr/testify/require"
+)
+
+// testCollectibleSend tests that we can properly send a collectible asset
+// with split commitments.
+func testCollectibleSend(t *harnessTest) {
+	// First, we'll make a collectible with emission enabled.
+	rpcAssets := mintAssetsConfirmBatch(
+		t, t.tarod, []*tarorpc.MintAssetRequest{issuableAssets[1]},
```
Comment (Roasbeef) [inline itest/collectible_split_test.go:15]: So if this is changed to `simpleAssets[1]` then everything works as expected?

File: itest/collectible_split_test.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	// Next, we'll attempt to complete three transfers of the full value of
+	// the asset between our main node and Bob.
+	var (
+		numSends     = 3
+		fullAmount   = rpcAssets[0].Amount
+		receiverAddr *tarorpc.Addr
+		err          error
+	)
+
+	for i := 0; i < numSends; i++ {
```
Comment (Roasbeef) [inline itest/collectible_split_test.go:42]: I think we can somewhat easily (factoring in the optional key family param) abstract this section into a sort of `pingPongAsset` helper func.

**Review by Roasbeef — COMMENTED** [review]
Set of changes looks pretty minimal which is great, will attempt to dig into the test failure when an asset w/ emission enabled is used to send. Might be some edge case in key family handling in the db, or just the coin selection portion itself.

**Comment by Roasbeef** [comment]
Confirmed, that the test works when using a collectible that doesn't support emission. 

Running the normal test I run into: 
```

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x1023cd454]

goroutine 1339 [running]:
github.com/lightninglabs/taro/mssmt.(*BranchNode).NodeHash(0x0?)
	/Users/roasbeef/gocode/src/github.com/lightninglabs/taro/mssmt/node.go:210 +0x24
github.com/lightninglabs/taro/commitment.(*TaroCommitment).TapLeaf(0x140011f1228)
	/Users/roasbeef/gocode/src/github.com/lightninglabs/taro/commitment/taro.go:156 +0x30
github.com/lightninglabs/taro/commitment.(*TaroCommitment).TapscriptRoot(0x7f17870c4797d810?, 0x0)
	/Users/roasbeef/gocode/src/github.com/lightninglabs/taro/commitment/taro.go:181 +0x2c
github.com/lightninglabs/taro/tarofreighter.(*ChainPorter).stateStep(0x140005bbf80, {0x9, {{0xdb, 0x3}, 0x14000634000}, {0x14000048780, 0x104012a80}, {{{0xa7, 0x74, 0x1, ...}, ...}, ...}, ...})
	/Users/roasbeef/gocode/src/github.com/lightninglabs/taro/tarofreighter/chain_porter.go:884 +0x1914
github.com/lightninglabs/taro/tarofreighter.(*ChainPorter).advanceStateUntil(0x140005bbf80, 0x140010f3ec8, 0xa)
	/Users/roasbeef/gocode/src/github.com/lightninglabs/taro/tarofreighter/chain_porter.go:493 +0x12c
github.com/lightninglabs/taro/tarofreighter.(*ChainPorter).taroPorter(0x140005bbf80)
	/Users/roasbeef/gocode/src/github.com/lightninglabs/taro/tarofreighter/chain_porter.go:226 +0x278
created by github.com/lightninglabs/taro/tarofreighter.(*ChainPorter).Start.func1
	/Users/roasbeef/gocode/src/github.com/lightninglabs/taro/tarofreighter/chain_porter.go:131 +0x284
```

File: asset/asset.go [hunk 7]
```diff
@@ -543,7 +543,13 @@ func (r *RawKeyGenesisSigner) SignGenesis(keyDesc keychain.KeyDescriptor,
 		return nil, nil, err
 	}
 
-	return tweakedPrivKey.PubKey(), sig, nil
+	// Since we'll never query lnd for a tweaked key, it doesn't matter if
+	// we lose the parity information here.
+	tweakedPubKey, _ := schnorr.ParsePubKey(
```
Comment (Roasbeef) [inline asset/asset.go:548]: I don't think we need this anymore after https://github.com/lightninglabs/taro/pull/187 ?

**Review by Roasbeef — APPROVED** [review]
LGTM 💈

Final comment is that I think we can safely drop forcing the family key to be even/odd given that we'll now always serialize the full version.

**Review by Roasbeef — APPROVED** [review]
LGTM 💈

Final comment is that I think we can safely drop forcing the family key to be even/odd given that we'll now always serialize the full version.

**Review by Roasbeef — APPROVED** [review]
LGTM 💈

Final comment is that I think we can safely drop forcing the family key to be even/odd given that we'll now always serialize the full version.

File: vm/vm_test.go [hunk 8]
```diff
@@ -362,6 +362,56 @@ func splitFullValueStateTransition(t *testing.T, validRootLocator,
 	}
 }
 
+func splitCollectibleStateTransition(t *testing.T,
```
Comment (guggero) [inline vm/vm_test.go:365]: nit: the outer `t` isn't needed. But I see that it's there in the other test generator functions too, so non blocking.

File: tarofreighter/chain_porter.go (hunk 4 above)
Comment (guggero) [reply tarofreighter/chain_porter.go:711]: Looks like `SendStatePreparedComplete` still exists in `parcel.go`.

File: tarofreighter/chain_porter.go [hunk 9]
```diff
@@ -600,6 +600,16 @@ func (p *ChainPorter) stateStep(currentPkg sendPackage) (*sendPackage, error) {
 		// to complete the send w/o merging inputs.
 		assetInput := elgigibleCommitments[0]
 
+		// If the key found for the input UTXO is not from the Taro
+		// keyfamily, something has gone wrong with the DB.
+		if assetInput.InternalKey.Family != tarogarden.TaroKeyFamily {
```
Comment (guggero) [inline tarofreighter/chain_porter.go:605]: Did we run into this somewhere? Or what's the reason for this specific check? Mostly just curious.

**Review by guggero — APPROVED** [review]
Very nice and clean, LGTM :tada:

File: tarofreighter/chain_porter.go (hunk 9 above)
Comment (Roasbeef) [reply tarofreighter/chain_porter.go:605]: I think was added during the debugging phase as a sort of sanity check to help track down the bug.
